extensions/parquet/table/src/main/java/io/deephaven/parquet/table/transfer/PrimitiveArrayAndVectorTransfer.java

```diff
@@ -0,0 +1,63 @@
+/**
+ * Copyright (c) 2016-2023 Deephaven Data Labs and Patent Pending
+ */
+package io.deephaven.parquet.table.transfer;
+
+import io.deephaven.base.verify.Assert;
+import io.deephaven.engine.rowset.RowSequence;
+import io.deephaven.engine.table.ColumnSource;
+import org.jetbrains.annotations.NotNull;
+
+import java.nio.Buffer;
+
+/**
+ * Used as a base class of transfer objects for arrays/vectors of primitive types.
+ */
+abstract class PrimitiveArrayAndVectorTransfer<T, E, B extends Buffer> extends ArrayAndVectorTransfer<T, E, B> {
+    PrimitiveArrayAndVectorTransfer(@NotNull final ColumnSource<?> columnSource, @NotNull final RowSequence tableRowSet,
+            final int maxValuesPerPage, final int targetPageSize, @NotNull final B buffer) {
+        super(columnSource, tableRowSet, maxValuesPerPage, targetPageSize, buffer);
+    }
+
+    @Override
+    public int transferOnePageToBuffer() {
+        // Clear any old buffered data
+        buffer.clear();
+        repeatCounts.clear();
+        // Fill the buffer with data from the table
+        transferOnePageToBufferHelper();
+        // Prepare buffer for reading
+        buffer.flip();
+        repeatCounts.flip();
+        return buffer.limit();
+    }
+
+    final boolean addEncodedDataToBuffer(@NotNull final EncodedData<E> data, boolean force) {
+        if (!repeatCounts.hasRemaining()) {
+            Assert.eqFalse(force, "force");
+            return false;
+        }
+        if (data.numValues > buffer.remaining()) {
```

Comment:
I don't think you should re-size the buffers down, but I do think you should refuse to write data that doesn't fit into maxValuesPerPage unless forced. Applies to all similar codecs.